hypothesis-python/src/hypothesis/_settings.py

```diff
@@ -317,10 +317,18 @@ def __repr__(self):
         bits = []
         for name, setting in all_settings.items():
             value = getattr(self, name)
-            if value != setting.default or not setting.hide_repr:
+            if value != setting.default and not setting.hide_repr:
                 bits.append('%s=%r' % (name, value))
         return 'settings(%s)' % ', '.join(sorted(bits))
 
+    def __str__(self):
```

Comment:
This is a reasonable method, but it should not be the `__str__` of the class.  Perhaps `_show_changed`?

If you want to keep this to non-default settings that's fine by me, but there are a few subtleties:

- drop the `and not setting.hide_repr` entirely - if a deprecated or hidden setting is changed it should  show up here
- handle the empty string specially in the header logic (in my suggested message, e.g. by omitting `->`)
- the repr for the database setting can be very long - consider truncating it or replacing it with just the class name and elipses, e.g. `database=DirectoryBasedExampleDatabase(...)` is already long enough, or even `database=...`.

Reply:
Seems to me like the path of the database is the most important part of that setting, right? It is very long, but it seems a shame to replace it with elipses.

Reply:
I was also thinking of `InMemoryDataBase({})` where the whole contents of the DB are part of the repr, but I see your point and the contents *should* be empty at this point.  So I'll consider this 'considered', whatever you decide to do 😄